### PR TITLE
Jgfouca/acme merge 06 01 2017

### PR DESCRIPTION
**Pull request by jgfouca**
Last 3 months of ACME work.

Test suite: scripts_regression_tests
Test baseline: 
Test namelist changes: 
Test status: [bit for bit, roundoff, climate changing]

Fixes [CIME Github issue #]

User interface changes?: 

Code review:

### CODE REVIEW

**Comment by jgfouca** [comment]
I will try to annotate tomorrow.

File: src/components/data_comps/docn/cime_config/config_component.xml [hunk 1]
```diff
@@ -67,13 +67,13 @@
     appropriate and is provided for testing and development purposes only.
     Users must create scientifically appropriate data for their particular
     application.  A tool is available to derive valid SOM forcing.
-    --- A setting of sst_aquapN (where N is an integer index value) is a 
-    type of prescribed SST mode used specifically for an aquaplanet setup in 
-    which global SSTs correspond to an analytic form set by the index value.  
+    --- A setting of sst_aquapN (where “N” is an integer index value) is a
```
Comment (mvertens) [inline src/components/data_comps/docn/cime_config/config_component.xml]: Please remove the "" here and below- this did not translate well in the $CASEROOT file. The quotes were removed on master - why did they get reintroduced in your PR. Did you rebase?

File: src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+
+<?xml-stylesheet type="text/xsl" href="namelist_defaults.xsl"?>
+
+<namelist_defaults>
+
+<drv_flds_in_files>Buildconf/camconf/drv_flds_in,Buildconf/clmconf/drv_flds_in</drv_flds_in_files>
+
+</namelist_defaults>
```
Comment (mvertens) [inline src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml]: There should be no namelists_defaults.xml file. These values are contained in namelist_definition_drv.xml. If ALM is using this, it is incorrect. The prognostic components should not be using xml files in the driver cime_config/ directory. For CESM, CLM was modified (with just some slight changes) to not use this.

File: src/drivers/mct/bld/namelist_files/namelist_definition_drv.xml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+       which has not been converted to parse the version 2.0 schema for
+       a namelist_definition.xml file
+  -->
+
+  <entry id="start_type" type="char*64" category="expdef"   group="seq_infodata_inparm" valid_values="startup,branch,continue"> </entry>
+  <entry id="start_ymd"  type="integer" category="time"     group="seq_timemgr_inparm"> </entry>
+  <entry id="glc_nec"    type="integer" category="seq_flds" group="seq_cplflds_inparm"> </entry>
+
+</namelist_definition>
+
```
Comment (mvertens) [inline src/drivers/mct/bld/namelist_files/namelist_definition_drv.xml]: This section makes no sense to me. First these entry_id values do not have the right schema - and they do not belong in the namelist_definition_drv.xml.

File: src/drivers/mct/bld/namelist_files/namelist_definition_drv_flds.xml [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  <!-- ========================================================================================  -->
+  <!-- CARMA fields                                                                              -->
+  <!-- ========================================================================================  -->
+
+  <entry id="carma_fields" type="char*256" category="carma"
+	 group="carma_inparm" valid_values="" >
+    List of fluxes needed by the CARMA model, from CLM to CAM.
+  </entry>
+
+</namelist_definition>
```
Comment (mvertens) [inline src/drivers/mct/bld/namelist_files/namelist_definition_drv_flds.xml]: The above has the wrong schema for xml driver cime_config/ files. These entries are already in namelist_definition_drv_flds.xml and should not be replicated here. If ALM needs these fields - they should be in the ALM bld/ directory (and the schema above is consistent with having them in the ALM bld/ directory) and not the driver cime_config/ directory.

File: src/drivers/mct/bld/testdir/env_build.xml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    Must be explicitly set in env_build.xml for userdefined machine.</desc>
+    </entry>
+    <entry id="PIO_VERSION" value="1">
+      <type>integer</type>
+      <valid_values>1,2</valid_values>
+      <desc>PIO library version</desc>
+    </entry>
+
+  </group>
+</file>
```
Comment (mvertens) [inline src/drivers/mct/bld/testdir/env_build.xml]: The above also has the wrong schema for an entry_id. Who is using this? Is it needed?

File: src/drivers/mct/bld/testdir/env_case.xml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+      <desc>case name</desc>
+    </entry>
+  </group>
+  <group id="case_desc">
+    <entry id="USER" value="UNSET">
+      <type>char</type>
+      <desc>case user name</desc>
+    </entry>
+  </group>
+</file>
```
Comment (mvertens) [inline src/drivers/mct/bld/testdir/env_case.xml]: See above comment for env_build.xml.

File: src/drivers/mct/bld/testdir/env_mach_pes.xml [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+      <valid_values>1</valid_values>
+      <desc>stride of mpi tasks for wav comp - currently should always be set to 1</desc>
+    </entry>
+    <entry id="PSTRID_ESP" value="1">
+      <type>integer</type>
+      <valid_values>1</valid_values>
+      <desc>stride of mpi tasks for esp comp - currently should always be set to 1</desc>
+    </entry>
+  </group>
+</file>
```
Comment (mvertens) [inline src/drivers/mct/bld/testdir/env_mach_pes.xml]: see above comment for env_build.xml - this is the wrong schema for the driver xml files and seems like a historical file that is no longer needed.

File: config/acme/machines/template.st_archive [hunk 8]
```diff
@@ -48,12 +48,36 @@ formatter_class=argparse.ArgumentDefaultsHelpFormatter
...
+
+    parser.add_argument("--copy-only", default=False, action="store_true",
+                        help="Copy instead of move the files to be archived")
+
+    parser.add_argument("--force-move", default=False, action="store_true",
+                        help="Move the files even if it's unsafe to do so")
+
+    args = parser.parse_args()
+
+    CIME.utils.handle_standard_logging_options(args)
```
Comment (jgfouca) [inline config/acme/machines/template.st_archive]: I think this needs to be changed to parse_args_and_handle_standard_logging_options

File: config/cesm/machines/template.st_archive [hunk 9]
```diff
@@ -47,12 +47,20 @@ formatter_class=argparse.ArgumentDefaultsHelpFormatter
     parser.add_argument("--caseroot", default=os.getcwd(),
                         help="Case directory to build")
 
-    args = CIME.utils.parse_args_and_handle_standard_logging_options(args, parser)
+    parser.add_argument("--no-incomplete-logs", default=False, action="store_true",
+                        help="Whether to archive logs which have been completed or not")
+
+    parser.add_argument("--copy-only", default=False, action="store_true",
```
Comment (jgfouca) [inline config/cesm/machines/template.st_archive]: I'm wondering why there are changes to CESM's st_archive?

File: scripts/Testing/Testcases/ICP_script [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+# Test setup:
+#======================================================================
+
+cd $CASEROOT
+
+touch cice_perf.out
+
+./xmlchange --file env_run.xml   --id CONTINUE_RUN      --val FALSE
+./xmlchange --file env_run.xml   --id REST_OPTION       --val none
+./xmlchange --file env_build.xml --id CICE_AUTO_DECOMP  --val false
```
Comment (jgfouca) [inline scripts/Testing/Testcases/ICP_script]: I'm quite sure these shell-based testcases need to be removed.

File: scripts/Testing/testreporter.pl [hunk 11]
```diff
@@ -0,0 +1,678 @@
+#!/usr/bin/env perl
+use Getopt::Long;
+use Data::Dumper;
+use LWP;
+use HTTP::Request;
+use HTTP::Request::Common qw(POST);
+use XML::LibXML;
```
Comment (jgfouca) [inline scripts/Testing/testreporter.pl]: Needs to be deleted.

File: scripts/lib/CIME/XML/env_mach_specific.py [hunk 12]
```diff
@@ -91,9 +91,9 @@ def load_modules(self, modules_to_load):
         if (module_system == "module"):
             self._load_module_modules(modules_to_load)
         elif (module_system == "soft"):
-            self._load_soft_modules(modules_to_load)
-        elif (module_system == "dotkit"):
-            self._load_dotkit_modules(modules_to_load)
+            self._load_modules_generic(modules_to_load)
```
Comment (jgfouca) [inline scripts/lib/CIME/XML/env_mach_specific.py]: Encapsulate environment systems that do not have a python interface into "generic".

File: scripts/lib/CIME/check_input_data.py [hunk 13]
```diff
@@ -126,6 +126,13 @@ def check_input_data(case, svn_loc=None, input_data_root=None, data_list_dir="Bu
     """
     # Fill in defaults as needed
     svn_loc = SVN_LOCS[get_model()] if svn_loc is None else svn_loc
+    err = run_cmd("svn --non-interactive --trust-server-cert ls %s" % svn_loc)[0]
+    expect(err == 0,
+"""
+Could not connect to svn repo '%s'
```
Comment (jgfouca) [inline scripts/lib/CIME/check_input_data.py]: Needs to be change to reflect recent fix that was done on ACME side.

File: src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml (hunk 2 above)
Comment (jgfouca) [inline src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml]: @rljacob do you know what happened here on the ACME side?

File: src/externals/pio2/src/clib/ncparser.pl [hunk 14]
```diff
@@ -0,0 +1,362 @@
+#!/usr/bin/perl 
+use strict;
+
+my $netcdf_incdir = $ARGV[0];
+my $pnetcdf_incdir = $ARGV[1];
+
```
Comment (jgfouca) [inline src/externals/pio2/src/clib/ncparser.pl]: @jayeshkrishna are all these PIO changes expected?

File: tools/mapping/gen_mapping_files/runoff_to_ocn/runoff_map.nml [hunk 15]
```diff
@@ -0,0 +1,15 @@
+  &input_nml
+   gridtype     = 'rtm'
+   file_roff    = '/glade/p/cesm/cseg/inputdata/lnd/clm2/rtmdata/rdirc.05.061026'
+   file_ocn     = '/glade/p/cesm/cseg/mapping/grids/gx3v7_120309.nc'
+   file_nn      = 'map_r05_to_gx3v7_nn_130507.nc '
+   file_smooth  = 'map_gx3v7_to_gx3v7_sm_e1000r300_130507.nc '
```
Comment (jgfouca) [inline tools/mapping/gen_mapping_files/runoff_to_ocn/runoff_map.nml]: What's the source of these tools/mapping changes? I don't see any recent changes on the ACME side,

Comment (rljacob) [inline tools/mapping/gen_mapping_files/runoff_to_ocn/runoff_map.nml]: @jonbob might know.

File: src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml (hunk 2 above)
Comment (rljacob) [inline src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml]: No.  Maybe @bishtgautam knows.

Comment (bishtgautam) [inline src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml]: I don't believe ALM is using this file. The [d0be438](https://github.com/ACME-Climate/ACME/commit/d0be438b6b3822d8456c7512d7cf86bc2174339a) in ACME by @agsalin moved `cime/driver_cpl/bld/namelist_files/namelist_defaults_drv.xml` to `src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml`.

File: config/cesm/machines/template.st_archive (hunk 9 above)
Comment (jgfouca) [inline config/cesm/machines/template.st_archive]: Perhaps someone from CESM can chime in to say if they're interested in these new st_archive features.

File: src/components/data_comps/docn/cime_config/config_component.xml (hunk 1 above)
Comment (jgfouca) [inline src/components/data_comps/docn/cime_config/config_component.xml]: I think removal of trailing whitespace caused a conflict.

File: src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml [hunk 16]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+
+<?xml-stylesheet type="text/xsl" href="namelist_defaults.xsl"?>
+
```
Comment (jgfouca) [inline src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/namelist_files/namelist_definition_drv.xml [hunk 17]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
+
+<namelist_definition>
+
```
Comment (jgfouca) [inline src/drivers/mct/bld/namelist_files/namelist_definition_drv.xml]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/namelist_files/namelist_definition_drv_flds.xml [hunk 18]
```diff
@@ -0,0 +1,131 @@
+<?xml version="1.0"?>
+
+<?xml-stylesheet type="text/xsl" href="namelist_definition.xsl"?>
+
+<namelist_definition version="1.0">
+
+  <entry id="drv_flds_in_files"
+	 type="char*256(10)"
```
Comment (jgfouca) [inline src/drivers/mct/bld/namelist_files/namelist_definition_drv_flds.xml]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/namelist_files/namelist_definition_modio.xml [hunk 19]
```diff
@@ -0,0 +1,122 @@
+<?xml version="1.0"?>
+
+<?xml-stylesheet type="text/xsl" href="http://www.cgd.ucar.edu/~cam/namelist/namelist_definition.xsl"?>
+
+<namelist_definition>
+
```
Comment (jgfouca) [inline src/drivers/mct/bld/namelist_files/namelist_definition_modio.xml]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/testdir/env_build.xml [hunk 20]
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0"?>
+<file id="env_build.xml">
+  <header>
+    These variables SHOULD NOT be changed once the model has been built.
+    urrently, these variables are not cached.
+    Note1: users SHOULD NOT modify BUILD_COMPETE below
+    this is done automatically by the scripts.
+    </header>
```
Comment (jgfouca) [inline src/drivers/mct/bld/testdir/env_build.xml]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/testdir/Buildconf/camconf/drv_flds_in [hunk 21]
```diff
@@ -0,0 +1,9 @@
+&drydep_inparm
+/
+&megan_emis_nl
```
Comment (jgfouca) [inline src/drivers/mct/bld/testdir/Buildconf/camconf/drv_flds_in]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/testdir/Buildconf/clmconf/drv_flds_in [hunk 22]
```diff
@@ -0,0 +1,9 @@
+&drydep_inparm
+/
+&megan_emis_nl
+/
```
Comment (jgfouca) [inline src/drivers/mct/bld/testdir/Buildconf/clmconf/drv_flds_in]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/testdir/env_case.xml [hunk 23]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<file id="env_case.xml">
+  <header>
+    These variables CANNOT BE CHANGED once a case has been created.
+    Invoke create_newcase again if a different grid or component
+    combination is required.
+    </header>
+  <group id="case_def">
```
Comment (jgfouca) [inline src/drivers/mct/bld/testdir/env_case.xml]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/testdir/env_mach_pes.xml [hunk 24]
```diff
@@ -0,0 +1,324 @@
+<?xml version="1.0"?>
+<file id="env_mach_pes.xml">
+  <header>
+    These variables CANNOT be modified once case_setup has been
+    invoked without first invoking case_setup -clean.
+
+    component task/thread settings
+    if the user wants to change the values below after ./case_setup, run
+    ./case_setup -clean
+    ./case_setup
```
Comment (jgfouca) [inline src/drivers/mct/bld/testdir/env_mach_pes.xml]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/testdir/env_run.xml [hunk 25]
```diff
@@ -0,0 +1,1278 @@
+<?xml version="1.0"?>
+<file id="env_run.xml">
+  <header>
+    These variables MAY BE CHANGED ANYTIME during a run.
+    Additional machine speific variables that can be changed
+    during a run are contained in the env_mach_specific file
+    Note1: users SHOULD NOT modify BUILD_COMPETE in env_build.xml
```
Comment (jgfouca) [inline src/drivers/mct/bld/testdir/env_run.xml]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/testdir/runtest [hunk 26]
```diff
@@ -0,0 +1 @@
+../build-namelist -caseroot . -cimeroot ../../.. -v -infile user_nl_cpl
```
Comment (jgfouca) [inline src/drivers/mct/bld/testdir/runtest]: Red flag. Addition of entirely new XML file.

Comment (jgfouca) [inline src/drivers/mct/bld/testdir/runtest]: Red flag. Addition of entirely new XML file.

File: src/drivers/mct/bld/testdir/user_nl_cpl [hunk 27]
```diff
@@ -0,0 +1,3 @@
+&a
+  drydep_method = 'xactive_lnd'
```
Comment (jgfouca) [inline src/drivers/mct/bld/testdir/user_nl_cpl]: Red flag. Addition of entirely new XML file.

File: src/externals/mct/mct/m_GlobalSegMap.F90 [hunk 28]
```diff
@@ -1764,22 +1761,25 @@ subroutine active_pes_(GSMap, n_active, pe_list)
   character(len=*),parameter :: myname_=myname//'::active_pes_'
 
   integer :: count, i, n, ngseg, ierr
-  logical :: new
-  integer, dimension(:), allocatable :: temp_list
-  integer, dimension(:), allocatable :: perm
+  integer :: max_activepe, p
```
Comment (jgfouca) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: @rljacob Can you review these MCT fortran changes?

File: tools/mapping/gen_mapping_files/runoff_to_ocn/runoff_map.nml [hunk 29]
```diff
@@ -0,0 +1,15 @@
...
+   file_new     = 'map_r05_to_gx3v7_nnsm_e1000r300_130507.nc'
+   title        = 'runoff map: r05 -> gx3v7, nearest neighbor and smoothed '
+   eFold        = 1000000.0
+   rMax         =  300000.0
+   step1 = .true.
```
Comment (jgfouca) [inline tools/mapping/gen_mapping_files/runoff_to_ocn/runoff_map.nml]: Red flag. Addition of entirely new file.

File: src/externals/mct/mct/m_GlobalSegMap.F90 (hunk 28 above)
Comment (rljacob) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: These are all from the MCT upgrade I mentioned.

**Comment by jgfouca** [comment]
Tests are passing. Still need to investigate PR's creation of new files which is probably just a merge problem.

File: tools/mapping/gen_mapping_files/runoff_to_ocn/runoff_map.nml (hunk 29 above)
Comment (rljacob) [inline tools/mapping/gen_mapping_files/runoff_to_ocn/runoff_map.nml]: "glade" is an NCAR thing so ACME didn't add this.  Is something that was deleted on ESMCI and this merge is trying to add it back?

File: src/externals/mct/mct/m_GlobalSegMap.F90 (hunk 28 above)
Comment (jgfouca) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: @rljacob there was a conflict in 
src/externals/mct/mpi-serial/tests/ftest_old.F90

so maybe that's the only file that needs a review.

File: src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml (hunk 2 above)
Comment (mvertens) [inline src/drivers/mct/bld/namelist_files/namelist_defaults_drv.xml]: There  is **not** be a drivers/mct/bld directory at this point - this needs to be removed. All of the files there have older schemas that are no longer used and were removed when the driver namelist was pythonized. The old bld/ directory was removed as part of the pythonization and everything needed to configure and build the driver is in its cime_config/ directory.

Also - the bld/namelist_defaults_drv.xml was refactored to the new python entry_id schema and is now the file drivers/mct/cime_config/namelist_definition_drv_flds.xml

**Comment by mvertens** [comment]
@jgfouca - the drivers/mct/bld directory needs to be removed- these are not new files - they were removed when the driver namelist was pythonized. I am not sure why this directory is reappearing in this merge - but it should not be here.

File: config/cesm/machines/template.st_archive (hunk 9 above)
Comment (mfdeakin-sandia) [inline config/cesm/machines/template.st_archive]: @jedwards4b asked me to try and keep the template files the same. I've left the --force-move option out as it's unsafe.

File: src/externals/mct/mct/m_GlobalSegMap.F90 (hunk 28 above)
Comment (rljacob) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: That file hasn't been touched since 2014.   This branch is fubar.

Comment (rljacob) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: Hasn't been touched in the mpi-serial side.  But in CIME, all the indentation was changed in f4a104b15

File: scripts/lib/CIME/provenance.py [hunk 30]
```diff
@@ -116,29 +114,40 @@ def save_prerun_provenance_acme(case, lid=None):
...
-                              ("xtnodestat >", "xtnodestat"),
-                              ("showq >", "showq")]:
-            full_cmd = cmd + " " + filename
-            run_cmd_no_fail(full_cmd + "." + lid, from_dir=full_timing_dir)
-            gzip_existing_file(os.path.join(full_timing_dir, filename + "." + lid))
-
-        mdiag_reduce = os.path.join(full_timing_dir, "mdiag_reduce." + lid)
-        run_cmd_no_fail("./mdiag_reduce.csh", arg_stdout=mdiag_reduce, from_dir=os.path.join(caseroot, "Tools"))
-        gzip_existing_file(mdiag_reduce)
+    if job_id is not None:
```
Comment (jedwards4b) [inline scripts/lib/CIME/provenance.py]: Hard coding based on machine name that is actually a BATCH_SYSTEM dependance.

File: scripts/lib/CIME/provenance.py [hunk 31]
```diff
@@ -250,32 +275,23 @@ def save_postrun_provenance_acme(case, lid):
...
     #
     globs_to_copy = []
-    if mach == "titan":
-        globs_to_copy.append("{}*OU".format(job_id))
-    elif mach == "mira":
-        globs_to_copy.append("{}*output".format(job_id))
-        globs_to_copy.append("{}*cobaltlog".format(job_id))
-    elif mach in ["edison", "cori-haswell", "cori-knl"]:
-        globs_to_copy.append("{}".format(case.get_value("CASE")))
+    if job_id is not None:
```
Comment (jedwards4b) [inline scripts/lib/CIME/provenance.py]: Do we really need machine specific code here?

File: src/drivers/mct/bld/testdir/user_nl_cpl (hunk 27 above)
Comment (jedwards4b) [inline src/drivers/mct/bld/testdir/user_nl_cpl]: I don't think that we want a testdir directory and if we do it should probably be located in scripts/tests

File: src/drivers/mct/cime_config/config_component.xml [hunk 32]
```diff
@@ -506,12 +506,12 @@
   <entry id="DOUT_S_SAVE_INTERIM_RESTART_FILES">
```
Comment (jedwards4b) [inline src/drivers/mct/cime_config/config_component.xml]: Maybe we should move this to the config_component_acme.xml file so that cesm can maintain the original default value of FALSE?

File: src/drivers/mct/cime_config/config_component.xml [hunk 33]
```diff
@@ -542,7 +542,7 @@
   <entry id="BATCH_SYSTEM">
      <type>char</type>
      <default_value>none</default_value>
-     <valid_values>moab,pbs,lsf,slurm,cobalt,cobalt_theta,none</valid_values>
+     <valid_values>lc_slurm,moab,pbs,lsf,slurm,cobalt,cobalt_theta,none</valid_values>
```
Comment (jedwards4b) [inline src/drivers/mct/cime_config/config_component.xml]: Why do we need lc_slurm?   Was sufficient effort made to use the slurm BATCH_SYSTEM definition?

File: src/share/util/mct_mod.F90 [hunk 34]
```diff
@@ -98,6 +98,7 @@ module mct_mod
    use m_GlobalSegMap       ,only: mct_gsMap_gstorage     => GlobalStorage
    use m_GlobalSegMap       ,only: mct_gsMap_ngseg        => ngseg
    use m_GlobalSegMap       ,only: mct_gsMap_nlseg        => nlseg
+   use m_GlobalSegMap       ,only: mct_gsMap_OP           => OrderedPoints
```
Comment (jedwards4b) [inline src/share/util/mct_mod.F90]: This is an error.  It requires a minor change in the clm component to remove it, but it should be removed and replaced with line 106.

File: src/share/util/shr_orb_mod.F90 [hunk 35]
```diff
@@ -78,8 +78,10 @@ real(SHR_KIND_R8) pure FUNCTION shr_orb_cosz(jday,lat,lon,declin,dt_avg)
       shr_orb_cosz =  shr_orb_avg_cosz(jday, lat, lon, declin, dt_avg)
    else
       shr_orb_cosz = sin(lat)*sin(declin) - &
-           cos(lat)*cos(declin) * &
-           cos((jday-floor(jday))*2.0_SHR_KIND_R8*pi + lon)
+      &              cos(lat)*cos(declin)*cos(jday*2.0_SHR_KIND_R8*pi + lon)
```
Comment (jedwards4b) [inline src/share/util/shr_orb_mod.F90]: We need the newer version here.

File: scripts/lib/CIME/case_st_archive.py [hunk 36]
```diff
@@ -149,6 +164,8 @@ def _archive_history_files(case, archive, archive_entry,
         for i in range(ninst):
             if compname == 'dart':
                 newsuffix = casename + suffix
+            elif compname.find('mpas') == 0:
```
Comment (jgfouca) [inline scripts/lib/CIME/case_st_archive.py]: @jedwards4b I don't know off the top of my head how I would do that.

File: scripts/lib/CIME/provenance.py [hunk 37]
```diff
@@ -16,21 +16,23 @@ def _get_batch_job_id_for_syslog(case):
...
-        return os.environ["PBS_JOBID"]
-    elif mach in ['edison', 'cori-haswell', 'cori-knl']:
-        return os.environ["SLURM_JOB_ID"]
-    elif mach == 'mira':
-        return os.environ["COBALT_JOBID"]
-    else:
-        return None
-
-def save_build_provenance_acme(case, lid=None):
+    try:
```
Comment (jgfouca) [inline scripts/lib/CIME/provenance.py]: It's complicated. I believe it has something to do with only wanting to do operations requiring job-id on certain machines. The ACME performance team is picky on maintaining exact compliance with the pre-CIME5 behavior.

File: scripts/lib/CIME/provenance.py (hunk 30 above)
Comment (jgfouca) [inline scripts/lib/CIME/provenance.py]: Same comment as above.

File: scripts/lib/CIME/provenance.py (hunk 31 above)
Comment (jgfouca) [inline scripts/lib/CIME/provenance.py]: Same comment as above. On the bright side, CESM doesn't need to worry about this code.

File: src/drivers/mct/cime_config/config_component.xml (hunk 33 above)
Comment (jgfouca) [inline src/drivers/mct/cime_config/config_component.xml]: Unfortunately, I have no idea. Just looking at ACME's lc_slurm XML, it's somewhat unique; looks like a cross between moab and slurm, so maybe this is OK.

File: src/drivers/mct/cime_config/config_component.xml (hunk 32 above)
Comment (jgfouca) [inline src/drivers/mct/cime_config/config_component.xml]: Good point.

File: src/externals/mct/mct/m_GlobalSegMap.F90 (hunk 28 above)
Comment (jgfouca) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: @rljacob so what should I do?

**Comment by jgfouca** [comment]
OK, I think I've address all the feedback, thanks everyone! I'm still unsure of mct and pio2 changes and so need further comments from @rljacob and @jayeshkrishna

File: src/share/util/mct_mod.F90 (hunk 34 above)
Comment (jgfouca) [inline src/share/util/mct_mod.F90]: @rljacob do you know what's going on here?

File: src/share/util/shr_orb_mod.F90 (hunk 35 above)
Comment (jgfouca) [inline src/share/util/shr_orb_mod.F90]: So you want to revert this change? @rljacob  that OK?

Comment (jgfouca) [inline src/share/util/shr_orb_mod.F90]: @rljacob Your commit made this change on ACME:

    Fix greather-then-roundoff diff in cos zen calc
    
    For jday larger than one, the argument (2*pi*jday + lon) for cos() will be
    reduced by the cos function, and as a result, some accuracy will be
    lost.  The reduction is different for different machines and results in
    greather-then-roundoff differences between machines for the same case.
    Solution is to change argument to ((jday-floor(jday))*2*pi + lon)

File: scripts/lib/CIME/provenance.py (hunk 37 above)
Comment (rljacob) [inline scripts/lib/CIME/provenance.py]: Yes the performance tracking queries the queue using the job-id.

File: src/share/util/shr_orb_mod.F90 (hunk 35 above)
Comment (jgfouca) [inline src/share/util/shr_orb_mod.F90]: Oh, it looks like we went back to the newer version on ACME.

File: src/share/util/mct_mod.F90 (hunk 34 above)
Comment (rljacob) [inline src/share/util/mct_mod.F90]: CLM should use the function name on line 106, not the one on 101.  Same function but different rename.

Comment (jgfouca) [inline src/share/util/mct_mod.F90]: The commit that made this change on the ACME side is:

commit 89940159774427904f816beca6e240fb35840dda
Author: Andy Salinger <agsalin@sandia.gov>
Date:   Fri Apr 14 11:57:23 2017 -0600

    Fixes to get tests to pass.
    
    scripts_regression_tests pass
    acme_developer has 1 remaining fail: HOMME_P24.f19_g16_rx1.A
    
    Changes:
    HOMME and cprnc CMakeFile fixes
    Undo mct_mod change that broke CLM build
    Remove unused <list>1<\list> in /clm/cime_config/config_component.xml
    Removed redundant MPASLI entries in env_run.xml
    Resolve conflict in update_acme_tests.py that was missed

File: src/externals/mct/mct/m_GlobalSegMap.F90 (hunk 28 above)
Comment (rljacob) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: @gold2718 why did you change this .F90 file?

File: scripts/lib/CIME/case_st_archive.py (hunk 36 above)
Comment (rljacob) [inline scripts/lib/CIME/case_st_archive.py]: This was the workaround that @mvertens came up with.

File: src/share/util/mct_mod.F90 (hunk 34 above)
Comment (rljacob) [inline src/share/util/mct_mod.F90]: So remove this line on this branch and when we next merge CIME to ACME, we'll make the required change in CLM.

File: src/share/util/shr_orb_mod.F90 (hunk 35 above)
Comment (rljacob) [inline src/share/util/shr_orb_mod.F90]: Yes we finally did but not before you started this merge.  So you can keep the ESMCI version of this file.

File: src/share/util/mct_mod.F90 (hunk 34 above)
Comment (rljacob) [inline src/share/util/mct_mod.F90]: What Andy should have done was change CLM.  Tagging @bishtgautam

File: src/drivers/mct/cime_config/config_component.xml (hunk 32 above)
Comment (rljacob) [inline src/drivers/mct/cime_config/config_component.xml]: Already have an issue for that (#1481)

File: src/externals/mct/mct/m_GlobalSegMap.F90 (hunk 28 above)
Comment (rljacob) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: @jgfouca if you're in a hurry, use the ESMCI version.

File: src/externals/pio2/src/clib/ncparser.pl (hunk 14 above)
Comment (jayeshkrishna) [inline src/externals/pio2/src/clib/ncparser.pl]: @jgfouca : I don't remember making any of these files on the ACME side. I suspect that the file was added in ACME to the new location when it was moved around ( ACME-Climate/ACME@d0be438b6b3822d8456c7512d7cf86bc2174339a) but was deleted in CIME (ncparser.pl is no longer available in the pio2 repo and CIME brought in a newer version of pio2)

Comment (jayeshkrishna) [inline src/externals/pio2/src/clib/ncparser.pl]: i.e., We don't need this file with the newer version of pio2

Comment (jayeshkrishna) [inline src/externals/pio2/src/clib/ncparser.pl]: Also, there is only one change made in pio2 directory (cime/src/externals/pio2) in the ACME side that we need to retain. The change was to rename the XML man pages (./pio2/examples/basic/perl5lib/XML/man3/XML_Lite_Element.3 , ./pio2/examples/basic/perl5lib/XML/man3/XML_Lite.3) to remove colons.

(The old filenames were ./pio2/examples/basic/perl5lib/XML/man3/XML::Lite.3 , ./pio2/examples/basic/perl5lib/XML/man3/XML::Lite::Element.3)

File: src/externals/mct/mct/m_GlobalSegMap.F90 (hunk 28 above)
Comment (gold2718) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: I never touched it. The changes you refer to were made by Pat in commit e1ea1d1e4dfd6deba93346e18a6c67a2fd38a89c in the MCT repo.
Perhaps it came into CIME in an MCT update pull request.

Comment (rljacob) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: This comment thread is attached to m_GlobalSegMap.F90 but the file I'm asking you about is src/externals/mct/mpi-serial/tests/ftest_old.F90

Comment (gold2718) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: I have never touched that file. It was affected by the move of mpi-serial to its own repo (before being brought into MCT as a subtree) but I don't think its contents have been altered in the MCT repo since 2014.

Comment (rljacob) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: git blame says you did in commit f4a104b1 in PR #1568 .

Its all white space except for changing the last line from "end" to "end program test".

@jgfouca you can take the ESMCI version.  Its better and I'll make the same change in mpi-serial.

Comment (gold2718) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: Of course my name is on that because it was a PR bringing in a new version of MCT as a CIME subtree. That doesn't mean I made the changes to MCT (or to mpi-serial which was brought into MCT as a new subtree). You need to look in those repositories for the true source of the changes. Also, note that you merged the PRs of mpi-serial into MCT.

File: src/drivers/mct/cime_config/config_component.xml (hunk 33 above)
Comment (rljacob) [inline src/drivers/mct/cime_config/config_component.xml]: @bogensch made those changes and can comment.

File: src/externals/mct/mct/m_GlobalSegMap.F90 (hunk 28 above)
Comment (rljacob) [inline src/externals/mct/mct/m_GlobalSegMap.F90]: Did you bring in some changes from @jedwards4b in that branch?  It was definitely one of you two.   ESMCI's master version is different from the current master in mpi-serial or MCT.  

Anyway we don't need to figure out "who" because the "why" appears to just be formatting.

**Comment by jgfouca** [comment]
I've went back over all the changes and comments one last time. I think this is ready. I've retested and everything looks good, so here goes...

**Comment by rljacob** [comment]
@jgfouca you should add "Fixes" to the description and also make a better title.
